tensorflow_addons/layers/optical_flow.py

```diff
@@ -0,0 +1,207 @@
+      stride_2: An integer specifying the stride length in the patch.
+      pad: An integer specifying the paddings in height and width.
+      data_format: Specifies the data format.
+          Possible values are:
+          "channels_last" float [batch, height, width, channels]
+          "channels_first" float [batch, channels, height, width]
+          Defaults to `"channels_last"`.
+      name: A name for the operation (optional).
+
+    Returns:
+      A `Tensor` of the format specified by `data_format`.
+    """
+
+    with tf.name_scope(name or "correlation_cost"):
+        op_call = _correlation_cost_op_so.correlation_cost
+
+        if data_format == "channels_last":
+            op_data_format = "NHWC"
+        elif data_format == "channels_first":
+            op_data_format = "NCHW"
+        else:
+            raise ValueError("`data_format` must be either `channels_last` or"
+                             "`channels_first`")
+
+        ret = op_call(
+            input_a,
+            input_b,
+            kernel_size=kernel_size,
+            max_displacement=max_displacement,
+            stride_1=stride_1,
+            stride_2=stride_2,
+            pad=pad,
+            data_format=op_data_format)
+        if data_format == 'channels_last':
+            # this is easier to maintain without
+            # specializing an additional cuda kernel
+            return tf.transpose(ret, [0, 2, 3, 1])
+        return ret
+
+
+@tf.RegisterGradient("CorrelationCost")
+def _correlation_cost_grad(op, grad_output):
+    kernel_size = op.get_attr("kernel_size")
+    max_displacement = op.get_attr("max_displacement")
+    stride_1 = op.get_attr("stride_1")
+    stride_2 = op.get_attr("stride_2")
+    pad = op.get_attr("pad")
+    data_format = op.get_attr("data_format")
+
+    input_a = tf.convert_to_tensor(op.inputs[0], name="input_a")
+    input_b = tf.convert_to_tensor(op.inputs[1], name="input_b")
+    grad_output_tensor = tf.convert_to_tensor(grad_output, name="grad_output")
+
+    op_call = _correlation_cost_op_so.correlation_cost_grad
+    grads = op_call(
+        input_a,
+        input_b,
+        grad_output_tensor,
+        kernel_size=kernel_size,
+        max_displacement=max_displacement,
+        stride_1=stride_1,
+        stride_2=stride_2,
+        pad=pad,
+        data_format=data_format)
+
+    grad_input_a = tf.convert_to_tensor(grads[0], name="grad_input_a")
+    grad_input_b = tf.convert_to_tensor(grads[1], name="grad_input_b")
+    return [grad_input_a, grad_input_b]
+
+
+@keras_utils.register_keras_custom_object
+class CorrelationCost(tf.python.keras.layers.Layer):
+    def __init__(self, kernel_size, max_displacement, stride_1, stride_2, pad,
+                 data_format, **kwargs):
+        self.kernel_size = kernel_size
+        self.max_displacement = max_displacement
+        self.stride_1 = stride_1
+        self.stride_2 = stride_2
+        self.pad = pad
+        self.data_format = data_format
```

Comment:
Is see no check for possible data-formats `channels_first`. Is this caught somewhere else?

Reply:
There was a check in the op, but I agree better to catch at layer construction.